### PR TITLE
Memo transactions

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Add support for memo transaction.
Pairs with changes in https://github.com/Concordium/concordium-ledger-app/pull/16.

## Changes

Allow sending transactions with memos. (simple, encrypted and scheduled)
Added memos to the transaction type and in the database.
Load transactions from the wallet-proxy's v1 path. (instead of v0)
Display memos in transaction view.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: app/utils/types.ts [hunk 1]
```diff
@@ -187,10 +189,15 @@ export enum TransactionKindId {
...
     Register_data = 21,
+    Simple_transfer_with_memo = 22,
+    Encrypted_transfer_with_memo = 23,
+    Transfer_with_schedule_and_memo = 24,
 }
+
 export interface SimpleTransferPayload {
     amount: string;
     toAddress: string;
+    memo?: string;
```
Comment (orhoj) [inline app/utils/types.ts:200]: The typing is better if you make a new interface:
```
export interface SimpleTransferWithMemoPayload extends SimpleTransferPayload {
    memo: string;
}
```
Leaving it as optional on the regular simple transfer is not correct as it cannot have a memo.

Comment (orhoj) [reply app/utils/types.ts:200]: And after writing this I saw `SimpleTransferWithMemo`, but then I don't understand why you need an optional `memo` here.

File: app/utils/types.ts [hunk 2]
```diff
@@ -786,7 +803,17 @@ export function instanceOfAccountTransactionWithSignature(
 export function instanceOfSimpleTransfer(
```
Comment (orhoj) [inline app/utils/types.ts:803]: I think the design is better if we keep the two transactions separate. So I would not expect `instanceOfSimpleTransfer` to match for one with a memo.

File: scripts/cbor.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable no-console */
+import cbor from 'cbor';
```
Comment (orhoj) [inline scripts/cbor.ts:2]: This file should be deleted.

File: package.json [hunk 4]
```diff
@@ -40,6 +40,7 @@
     "test-e2e": "node ./internals/scripts/CheckBuildsExist.js && cross-env NODE_ENV=test testcafe electron:./app ./test/e2e/HomePage.e2e.ts",
     "test-e2e-live": "node ./internals/scripts/CheckBuildsExist.js && cross-env NODE_ENV=test testcafe --live electron:./app ./test/e2e/HomePage.e2e.ts",
     "test-watch": "yarn test --watch",
+    "cbor": "ts-node --project scripts/migrations/tsconfig.json -r tsconfig-paths/register ./scripts/cbor.ts",
```
Comment (orhoj) [inline package.json:43]: Delete?

File: app/utils/transactionSerialization.ts [hunk 5]
```diff
@@ -34,28 +34,42 @@ import {
...
+    const amount = encodeWord64(BigInt(payload.amount));
+    return Buffer.concat([kind, address, amount]);
+}
 
-    serialized[0] = TransactionKind.Simple_transfer;
-    putBase58Check(serialized, 1, payload.toAddress);
-    put(serialized, 32 + 1, encodeWord64(BigInt(payload.amount)));
-    return Buffer.from(serialized);
+function serializeSimpleTransferWithMemo(payload: SimpleTransferPayload) {
+    if (!payload.memo) {
```
Comment (orhoj) [inline app/utils/transactionSerialization.ts:55]: It should not be necessary to have this check here if we type the input parameter better.

File: app/utils/transactionHelpers.ts [hunk 6]
```diff
@@ -592,13 +613,36 @@ export function validateBakerStake(
...
 }
 
+export function validateMemo(memo: string): string | undefined {
+    const asNumber = Number(memo);
+    if (
+        Number.isInteger(asNumber) &&
+        (asNumber > Number.MAX_SAFE_INTEGER ||
+            asNumber < Number.MIN_SAFE_INTEGER)
+    ) {
+        return `Only numbers up to ${Number.MAX_SAFE_INTEGER} is supported.`;
```
Comment (orhoj) [inline app/utils/transactionHelpers.ts:623]: I suggest replacing the above with:
```
Numbers greater than 9007199254740991 or smaller than -9007199254740991 are not supported.
```
If there is a space concern and you want to keep the current, then you need to change _is_ to _are_.

File: app/utils/transactionHelpers.ts [hunk 7]
```diff
@@ -592,13 +613,36 @@ export function validateBakerStake(
...
+    }
+    if (getEncodedSize(memo) > maxMemoSize) {
+        return `Memo is too large, encoded size must be at most ${maxMemoSize} bytes`;
+    }
+    // Check that the memo only contains ascii characters
+    // eslint-disable-next-line no-control-regex
+    if (/[^\u0000-\u007f]/.test(memo)) {
```
Comment (orhoj) [inline app/utils/transactionHelpers.ts:630]: This deserves its own utility function as it is a candidate for reuse. So refactor it into an `isASCII` method.

File: app/utils/transactionHandlers/SimpleTransferHandler.tsx [hunk 8]
```diff
@@ -55,6 +55,7 @@ export default class SimpleTransferHandler
         amount,
         recipient,
         signatureAmount,
+        memo,
```
Comment (orhoj) [inline app/utils/transactionHandlers/SimpleTransferHandler.tsx:58]: The memo transactions should, in my opinion, have their own separate transaction handlers.

File: app/utils/transactionCosts.ts [hunk 9]
```diff
@@ -75,7 +76,10 @@ export function calculateCost(
     );
 }
 
-function getPayloadSizeEstimate(transactionKind: TransactionKindId) {
+/**
+   Note that this function does not support scheduled transfers nor transfers with memos.
```
Comment (orhoj) [inline app/utils/transactionCosts.ts:80]: It would be good to document what the function does now that you have written what it _does  not_.

File: app/utils/transactionCosts.ts [hunk 10]
```diff
@@ -125,24 +131,38 @@ function getEnergyCostOfType(transactionKind: TransactionKindId) {
     }
 }
 
-export function getScheduledTransferPayloadSize(scheduleLength: number) {
+export function getScheduledTransferPayloadSize(
+    scheduleLength: number,
+    memoLength?: number
```
Comment (orhoj) [inline app/utils/transactionCosts.ts:136]: This is never called without `memoLength` being set.

File: app/utils/transactionCosts.ts [hunk 11]
```diff
@@ -125,24 +131,38 @@ function getEnergyCostOfType(transactionKind: TransactionKindId) {
...
+        scheduleLength * 16 +
+        // Memo
+        (memoLength ? 2 + memoLength : 0)
+    );
 }
 
 /**
  *  Given the signatureAmount and schedule length,
  * returns the energy cost of a scheduled transfer.
+ * Param memoLength
```
Comment (orhoj) [inline app/utils/transactionCosts.ts:155]: Use @param to document input parameters.

File: app/utils/cborHelper.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
+/* eslint-disable import/prefer-default-export */
```
Comment (orhoj) [inline app/utils/cborHelper.ts:1]: I don't think you need this eslint-disable anymore.

File: app/utils/cborHelper.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Given a string or number, return a buffer containing the value under cbor encoding.
+ * N.B. given a string, this will attempt to convert it to a number, to decrease the encoded size.
+ */
+export function encodeAsCBOR(value: string | number): Buffer {
+    // Prefer saving as numbers:
+    const asNumber = Number(value);
+    if (Number.isInteger(asNumber)) {
+        if (
+            asNumber >= Number.MAX_SAFE_INTEGER ||
```
Comment (orhoj) [inline app/utils/cborHelper.ts:14]: `Number.MAX_SAFE_INTEGER` is also safe, so the equality should be `>` and `<` respectively.

File: app/utils/cborHelper.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    return Buffer.from(encode(value));
+}
+
+/**
+ * Decode cbot encoding.
+ * @param value is assumed to be a Hex string, containing cbor encoded bytes.
+ */
+export function decodeCBOR(value: string) {
+    return decode(Buffer.from(value, 'hex'));
```
Comment (orhoj) [inline app/utils/cborHelper.ts:29]: This seems to return an `any` type, and elsewhere we `toString()` it. What happens if the `memo` we try to decode is not `CBOR`, or if it is not something that `toString()` works nicely on?

**Comment by orhoj** [comment]
Very large integer input for a memo crashes the application.

File: app/pages/multisig/AccountTransactions/CreateTransferProposal/CreateTransferProposal.tsx [hunk 15]
```diff
@@ -79,14 +82,20 @@ function CreateTransferProposal({
 }: Props): JSX.Element {
     const dispatch = useDispatch();
 
+    const consensusStatus = useConsensusStatus();
+    const allowMemo = consensusStatus && consensusStatus.protocolVersion > 1;
```
Comment (orhoj) [inline app/pages/multisig/AccountTransactions/CreateTransferProposal/CreateTransferProposal.tsx:86]: Refactor into a shared method (as this is used in multiple places) calling something like `nodeSupportsMemo()`.

File: app/pages/Accounts/TransactionListElement.tsx [hunk 16]
```diff
@@ -231,6 +275,7 @@ interface Props {
 function TransactionListElement({
     transaction,
     onClick,
+    showFullMessage = false,
```
Comment (orhoj) [inline app/pages/Accounts/TransactionListElement.tsx:278]: Should this be `showFullMemo`?

File: app/pages/Accounts/TransactionListElement.tsx [hunk 17]
```diff
@@ -214,6 +221,42 @@ function statusSymbol(status: TransactionStatus) {
     }
 }
 
+const memoTypes = [
+    TransactionKindString.Transfer,
```
Comment (orhoj) [inline app/pages/Accounts/TransactionListElement.tsx:225]: It is confusing that `Transfer`, `EncryptedAmountTransfer` and `TransferWithSchedule` are part of `memoTypes` as they are not memo types.

File: app/pages/Accounts/TransactionListElement.tsx [hunk 18]
```diff
@@ -214,6 +221,42 @@ function statusSymbol(status: TransactionStatus) {
...
+
+function showMemo(
+    memo: string | undefined,
+    showFullMessage: boolean,
+    transactionKind: TransactionKindString
+) {
+    if (!memo && showFullMessage && memoTypes.includes(transactionKind)) {
+        return (
+            <i className="body4 m0 mT5 textFaded">
+                The transaction contains no message
```
Comment (orhoj) [inline app/pages/Accounts/TransactionListElement.tsx:241]: `message` -> `memo`.

File: app/features/ledger/Transfer.ts [hunk 19]
```diff
@@ -57,6 +60,38 @@ const INS_REMOVE_BAKER = 0x14;
...
+type Memo = string;
+
+/**
+ * Send memo to ledger for signing.
+ * N.B. This does not send the length of the memo, and the CBOR encoded string.
+ */
+async function sendMemo(
+    transport: Transport,
+    ins: number,
+    memo: Memo | undefined,
```
Comment (orhoj) [inline app/features/ledger/Transfer.ts:72]: If we type thing correctly, then this should not allow for an undefined value as far as I see it.

File: app/features/ledger/Transfer.ts [hunk 20]
```diff
@@ -57,6 +60,38 @@ const INS_REMOVE_BAKER = 0x14;
...
+    if (!memo) {
+        throw new Error('Unexpected Missing memo');
+    }
+
+    const encodedMemo = encodeAsCBOR(memo);
+
+    let response;
+    const chunks = chunkBuffer(encodedMemo, 255);
+    for (const chunk of chunks) {
+        // eslint-disable-next-line  no-await-in-loop
```
Comment (orhoj) [inline app/features/ledger/Transfer.ts:85]: I thought we had disabled the check for `no-await-in-loop` globally?

File: app/features/ledger/Transfer.ts [hunk 21]
```diff
@@ -57,6 +60,38 @@ const INS_REMOVE_BAKER = 0x14;
...
+        response = await transport.send(0xe0, ins, p1, p2, Buffer.from(chunk));
+    }
+    if (!response) {
+        throw new Error('Unexpected missing response from ledger;');
+    }
+    const signature = response.slice(0, 64);
```
Comment (orhoj) [inline app/features/ledger/Transfer.ts:91]: Else you refactored this from
```
    const signature = response.slice(0, 64);
    return signature;
```
to
```
    return response.slice(0, 64);
```
Pick one and stay consistent.

File: jest.config.js [hunk 22]
```diff
@@ -10,7 +10,10 @@ module.exports = {
     },
     moduleFileExtensions: ['js', 'jsx', 'ts', 'tsx', 'json'],
     moduleDirectories: ['node_modules', 'app/node_modules'],
-    setupFiles: ['./internals/scripts/CheckBuildsExist.js'],
+    setupFiles: [
+        './internals/scripts/CheckBuildsExist.js',
+        './test/loadModules.ts',
```
Comment (orhoj) [inline jest.config.js:15]: Why is this needed?

File: app/utils/transactionCosts.ts (hunk 10 above)
Comment (shjortConcordium) [reply app/utils/transactionCosts.ts:136]: Some tests did, but I changed that.

File: app/utils/cborHelper.ts (hunk 14 above)
Comment (shjortConcordium) [reply app/utils/cborHelper.ts:29]: It seems that it crashes.

**Comment by shjortConcordium** [comment]
> Very large integer input for a memo crashes the application.

 Yes, it is because the encodedSize is calculated (to calculate cost) while inputting, and I made the encode function crash if given a very large integer.
 I have changed it getEncodedSize catches the error and returns 0 instead.

File: app/pages/Accounts/TransactionListElement.tsx (hunk 17 above)
Comment (shjortConcordium) [reply app/pages/Accounts/TransactionListElement.tsx:225]: I have refactored the variable away.

File: app/features/ledger/Transfer.ts (hunk 21 above)
Comment (shjortConcordium) [reply app/features/ledger/Transfer.ts:91]: I have refactored it for all Transfers.

File: jest.config.js (hunk 22 above)
Comment (shjortConcordium) [reply jest.config.js:15]: Because otherwise we get a ReferenceError on TextDecoder on test files that import the cbor dependency.

File: app/utils/types.ts (hunk 1 above)
Comment (shjortConcordium) [reply app/utils/types.ts:200]: I have refactored it to make the transaction more separate.

File: app/utils/cborHelper.ts (hunk 14 above)
Comment (shjortConcordium) [reply app/utils/cborHelper.ts:28]: I have handled this in TransactionConverters (the only place this is used). It tries to decode, and otherwise saves the Hex string directly.

File: app/utils/transactionHelpers.ts [hunk 23]
```diff
@@ -592,13 +665,35 @@ export function validateBakerStake(
...
 }
 
+export function validateMemo(memo: string): string | undefined {
+    const asNumber = Number(memo);
+    if (
+        Number.isInteger(asNumber) &&
+        (asNumber > Number.MAX_SAFE_INTEGER ||
+            asNumber < Number.MIN_SAFE_INTEGER)
+    ) {
+        return `Numbers greater than ${Number.MAX_SAFE_INTEGER} or smaller than ${Number.MIN_SAFE_INTEGER} are not supported..`;
```
Comment (orhoj) [inline app/utils/transactionHelpers.ts:675]: Too many `.`.

**Comment by orhoj** [comment]
When receiving a transaction with a memo the memo does not show.

**Comment by orhoj** [comment]
Receiving an encrypted transfer with a memo does not work correctly.

**Comment by orhoj** [comment]
After merge with the account recovery branch I get the following warning when recovering:
```
Each child in a list should have a unique "key" prop.

Check the render method of `DisplayRecovery`
```

**Review by orhoj — APPROVED** [review]
